File: Sorting/mergesort.c
#include <stdio.h>

void merge(int ar[], int start, int mid, int end){
    int i, p = start, q = mid+1, A[end-start+1];
    for(i = 0; i<end-start+1; i++){
        if(p>mid){
            A[i] = ar[q++];
        }else if(q>end){
            A[i] = ar[p++];
        }else if(ar[p]>ar[q]){
            A[i] = ar[q++];
        }else{
            A[i] = ar[p++];
        }
    }

    for(q = start, i = 0; q<=end; q++, i++){
        ar[q] = A[i];
    }

}

void mergesort(int ar[], int start, int end){
    if(start<end){
        int mid = (start+end)/2;
        mergesort(ar,start,mid);
        mergesort(ar,mid+1,end);
        merge(ar,start,mid,end);
    }
}

int main(){
     int size = 8, ar[] = {1,3,2,7,6,4,3,12};
    mergesort(ar, 0, size-1);
    int i;
    for(i = 0; i<size; i++)
        printf("%d ", ar[i]);

    return 0;


}

File: Sorting/Python/mergesort.py
def merge(x, start, mid, end):
    A = [0] * (end-start+1)
    i = start
    j = mid+1
    index = 0
    while i <= mid and j <= end:
        if x[i] <= x[j]:
            A[index] = x[i]
            i += 1
        else:
            A[index] = x[j]
            j += 1
        index += 1
    
    while i <= mid:
        A[index] = x[i]
        i += 1
        index += 1
    
    while j <= end:
        A[index] = x[j]
        j += 1
        index += 1

    index = 0
    for i in range(start, end+1):
        x[i] = A[index]
        index += 1
    
    return


def mergeSort(x, start, end):
    if start < end:
        mid = (start + end)/2
        mergeSort(x, start, mid)
        mergeSort(x, mid+1, end)
        merge(x, start, mid, end)

def main():
    testCaseOne = [5,4,3,2,1]
    testCaseTwo = [1,5,3,6,8,-1]

    mergeSort(testCaseOne, 0, len(testCaseOne)-1)
    mergeSort(testCaseTwo, 0, len(testCaseTwo)-1)

    print("Testcase 1: ", testCaseOne)
    print("Testcase 2: ", testCaseTwo)


if __name__ == "__main__":
    main()
